net/libfab/libfab_internal.h

```diff
@@ -0,0 +1,398 @@
+
+	/** Passive endpoint (listening/server mode) */
+	struct m0_fab__ep         *ftm_pep;
+	
+	/** Shared receive context for shared buffer pools */
+	struct fid_ep             *ftm_rctx;
+	
+	/** Transmit Completion Queue */
+	struct fid_cq             *ftm_tx_cq;
+	
+	/** Shutdown flag */
+	bool                       ftm_shutdown;
+
+	/** List of completed buffers */
+	struct m0_tl               ftm_done;
+	
+	/** Lock used betn poller & tm_fini during shutdown */
+	struct m0_mutex            ftm_endlock;
+	
+	/** Lock used betn poller & tm_fini for posting event */
+	struct m0_mutex            ftm_evpost;
+
+	/** List of pending bulk operations */
+	struct m0_tl               ftm_bulk;
+};
+
+/**
+ * Libfab structure of buffer memory region parameters
+ */
+struct m0_fab__buf_mr {
+	/** Local memory region (buffer) descriptor */
+	void          *bm_desc[FAB_IOV_MAX];
+	
+	/** Memory region registration */
+	struct fid_mr *bm_mr[FAB_IOV_MAX];
+};
+
+/**
+ * Libfab structure of buffer descriptor 
+ * sent from the passive side to the active side
+ */
+struct m0_fab__bdesc {
+	/** Remote buffer iov count */
+	uint64_t fbd_iov_cnt;
+
+	/** Remote node address */
+	uint64_t fbd_netaddr;
+	
+	/** Remote buffer address */
+	uint64_t fbd_bufptr;
+};
+
+/**
+ * Libfab structure of buffer params
+ */
+struct m0_fab__buf {
+	/** Magic number for list of completed buffers */
+	uint64_t               fb_magic;
+	
+	/** Magic number for list of send buffers */
+	uint64_t               fb_sndmagic;
+	
+	/** Dummy data + network buffer ptr */
+	uint64_t               fb_dummy[2];
+	
+	/** Buffer descriptor of the remote node */
+	struct m0_fab__bdesc  *fb_rbd;
+	
+	/** Array of iov for remote node  */
+	struct fi_rma_iov     *fb_riov;
+	
+	/** Buffer memory region registration params */
+	struct m0_fab__buf_mr  fb_mr;
+	
+	/** Domain to which the buffer is registered */
+	struct fid_domain     *fb_dp;
+	
+	/** Pointer network buffer structure */
+	struct m0_net_buffer  *fb_nb;
+	
+	/** endpoint associated with receive operation */
+	struct m0_fab__ep     *fb_ev_ep;
+	
+	/** Transmit Context to be returned in the buffer completion event */
+	struct m0_fab__ep     *fb_txctx;
+	
+	/** Link in list of completed buffers */
+	struct m0_tlink        fb_linkage;
+	
+	/** Link in list of send buffers */
+	struct m0_tlink        fb_snd_link;
+
+	/** Buffer completion status */
+	int32_t                fb_status;
+	
+	/** Total size of data to be received/sent/read/written */
+	m0_bindex_t            fb_length;
+	
+	/** Count of work request generated for bulk rma operation */
+	volatile uint32_t      fb_wr_cnt;
+
+	/** Count of work request completions for bulk rma operation */
+	volatile uint32_t      fb_wr_comp_cnt;
+
+	/** Pointer to the m0_fab__bulk_op structure */
+	void*                  fb_bulk_op;
+};
+
+/**
+ * Libfab structure of connection data
+ */
+struct m0_fab__conn_data {
+	/** address in network byte format */
+	uint64_t fcd_netaddr;
```

Comment:
Is this compatible for IPv4 and IPv6?

Reply:
Currently only IPv4 is supported. We will add support for IPv6 in the future.